### PR TITLE
fix(zod): correct inline reusable schema emission (missing import + dropped defs)

### PR DESCRIPTION
**Pull request by z4o4z**
## Summary

Two pre-existing bugs in the **inline single-file** reusable-schema writer (`client: 'zod'` + `override.zod.generateReusableSchemas`, with no `schemas:` directory). Both produce output that does not compile. Found while working on #3467; fixed here independently.

### 1. Missing `import * as zod from 'zod'`

The zod client's import is a **usage-gated dependency import** — emitted only when an operation's generated schema references the `zod` token. When every operation is a pure-`$ref` alias, e.g.

```ts
export const GetThingResponse = Thing
```

the client emits no zod import. The inline schema block also skipped it (it keyed off `!hasOperations`), so the file ended up referencing `zod` with **no import at all**.

Now the inline block supplies the import unless an operation actually references `zod` (mirroring the client's own gate), so there is always exactly one zod import.

### 2. Dropped definitions for sanitized names

The inline writer seeded reusable refs from `builder.schemas`, whose `name` is the **sanitized** model identifier (`__schema0` → `_Schema0`). The resulting ref `#/components/schemas/_Schema0` doesn't exist in `components.schemas` (the real key is `__schema0`), so the definition was **silently dropped** whenever the schema was reachable only from operations — leaving the operation wrapper that references it dangling. (It survived when another component schema referenced it, via transitive expansion, which masked the bug.)

Now it seeds from the raw `components.schemas` keys, exactly like the per-file writer `writeZodSchemasReusable` already does (see its comment, added in #3465).

## Test plan

- [x] New unit test (`write-zod-specs`): `generateZodSchemasInline` emits a schema whose raw name needs sanitizing (`__my_data` → `_MyData`)
- [x] New e2e test (`generate-spec`): a spec whose only operation is a pure-`$ref` alias yields exactly one `from 'zod'` import (previously zero)
- [x] Existing inline test (operation that *does* use `zod`) still asserts exactly one import — no double-import regression
- [x] Generated output type-checks under `strict` on **zod v3 and v4** for both scenarios
- [x] `lint`, `typecheck`, and full `orval` suite (119 tests) pass

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **Bug Fixes**
  * Ensure Zod import is emitted only when generated operations require it (fixes missing/extra import for pure `$ref` responses).
  * Fix reusable-schema detection to correctly recognize component schemas from raw spec keys and eliminate leftover reference placeholders.

* **Tests**
  * Added regression tests for inline schema generation with pure `$ref` operations and correct Zod import behavior.
  * Added tests for component schema naming edge cases to prevent sentinel markers.

[Review Change Stack](https://app.coderabbit.ai/change-stack/orval-labs/orval/pull/3468?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: packages/orval/src/generate-spec.test.ts [hunk 1]
```diff
@@ -162,3 +162,75 @@ describe('generateSpec - generateReusableSchemas inline (single mode)', () => {
...
+      await generateSpec(workspace, options);
+
+      const content = await fs.readFile(targetFile, 'utf8');
+
+      // The component schema is defined inline...
+      expect(content).toContain('export const Thing = zod.object(');
+      // ...the operation wrapper is a pure-$ref alias (no `zod.` usage)...
+      expect(content).toContain('export const GetThingResponse = Thing');
+      // ...and exactly one zod import is present (the bug emitted zero).
+      expect(content.match(/from 'zod'/g) ?? []).toHaveLength(1);
```
[the remark this answers is not in the document]
Comment (z4o4z) [reply packages/orval/src/generate-spec.test.ts]: Done. Changed the assertion to `/from ['"]zod['"]/g` so it tolerates either quote style while still asserting exactly one `zod` module import. generate-spec.test.ts:230.

File: packages/orval/src/write-zod-specs.ts [hunk 2]
```diff
@@ -325,9 +322,21 @@ function generateZodSchemasInlineReusable(
...
+  // dropped whenever it was reachable only from operations (when referenced by
+  // another component schema it survived via transitive expansion, masking the
+  // bug). Mirrors `writeZodSchemasReusable`, which already seeds from raw keys.
+  const componentSchemas = (builder.spec.components?.schemas ?? {}) as Record<
+    string,
+    unknown
+  >;
+  const refs = Object.keys(componentSchemas).map(
+    (schemaName) => `#/components/schemas/${schemaName}`,
   );
```
[the remark this answers is not in the document]
Comment (z4o4z) [reply packages/orval/src/write-zod-specs.ts:397]: Leaving this as-is. The inline path deliberately mirrors the sibling `writeZodSchemasReusable` (write-zod-specs.ts:493-495), which also seeds from raw `Object.keys(components.schemas)` with no sort. More importantly, this list is not the emit order: the refs feed `generateReusableSchemaSet` -> `rewriteReusableSchemas`, where Tarjan's SCC (reusable-schemas.ts:315-356) reorders entries into reverse-topological order so dependencies emit before dependents. Within a single parse of a single spec, `Object.keys` order is stable, so output is reproducible. Adding a sort to only this path would diverge the two sibling paths' ordering for the same spec — the exact inconsistency the code comment at line 331 guards against. If deterministic key sorting is wanted it should be applied to both seeds in a dedicated change, not one-sided in this bugfix.

File: packages/orval/src/write-specs.ts [hunk 3]
```diff
@@ -450,10 +450,17 @@ export async function writeSpecs(
...
-    const includeZodImport = !hasOperations;
+    // The zod client's `import * as zod from 'zod'` is a *usage-gated* dependency
+    // import: it's only emitted when an operation's generated schema actually
+    // references the `zod` token. When every operation is a pure-`$ref` alias
+    // (e.g. `export const FooResponse = Bar`), the client emits no zod import —
+    // so the inline schema block (which always uses zod) must supply it itself.
+    // When an operation does use zod the client already imports it, and a second
+    // import would redeclare the `zod` binding — so the inline block omits it.
+    const operationsUseZod = Object.values(builder.operations).some(
+      (operation) => /\bzod\b/.test(operation.implementation),
```
[the remark this answers is not in the document]
Comment (z4o4z) [reply packages/orval/src/write-specs.ts:461]: Keeping `/\bzod\b/` here. This check exists to agree with the generator's OWN dependency-import gate: `addDependency` (core/src/generators/imports.ts:254-256) decides whether to emit `import * as zod from 'zod'` by matching `new RegExp(`\b(${searchWords})\b`)` (searchWords = `zod`) against the same `implementation` string, and single-mode.ts:88 does the same. The whole point is that this inline-import decision uses the identical heuristic on the identical string as the emission decision — so they can never disagree (no duplicate import, no missing one). Tightening to `\bzod\s*\.` would diverge from the generator's gate and reintroduce that risk. There is also no structured signal to use instead: the zod import comes from the global `ZOD_DEPENDENCIES` list (zod/src/index.ts:45-60), not from the operation; `operation.imports` only carries reusable schema `usedRefs` (index.ts:2073), and there is no `usesZod` flag. Adding one would mean refactoring the dependency mechanism — out of scope for this bugfix.

**Comment by melloware** [comment]
looks like Conflicts now

**Review by melloware — APPROVED** [review]
